clients/js/src/cmds/submit.ts

```diff
@@ -67,93 +79,150 @@ export const handler = async (
+    // or a token transfer VAA), in which case we just submit the VAA to
+    // that target chain.
+    //
+    // If the VAA does not have a target (e.g. chain registration VAAs or
+    // guardian set upgrade VAAs), we require the '--chain' argument to be
+    // set on the command line.
+    //
+    // As a sanity check, in the event that the VAA does specify a target
+    // and the '--chain' argument is also set, we issue an error if those
+    // two don't agree instead of silently taking the VAA's target chain.
+
+    // get chain from command line arg
+    const cli_chain = argv["chain"];
+
+    let chain: ChainName;
+    if (cli_chain !== undefined) {
+      assertChain(cli_chain);
+      if (vaa_chain !== "unset" && cli_chain !== vaa_chain) {
+        throw Error(
+          `Specified target chain (${cli_chain}) does not match VAA target chain (${vaa_chain})`
+        );
+      }
+      chain = coalesceChainName(cli_chain);
+    } else {
+      chain = vaa_chain;
+    }
+
+    chains.push(chain);
   }
 
-  if (chain === "unset") {
-    throw Error(
-      "This VAA does not specify the target chain, please provide it by hand using the '--chain' flag."
-    );
-  } else if (isEVMChain(chain)) {
-    await execute_evm(
-      parsed_vaa.payload,
-      buf,
-      network,
-      chain,
-      argv["contract-address"],
-      argv["rpc"]
-    );
-  } else if (isTerraChain(chain)) {
-    await execute_terra(parsed_vaa.payload, buf, network, chain);
-  } else if (chain === "solana" || chain === "pythnet") {
-    await execute_solana(parsed_vaa, buf, network, chain);
-  } else if (chain === "algorand") {
-    await execute_algorand(
-      parsed_vaa.payload,
-      new Uint8Array(Buffer.from(vaa_hex, "hex")),
-      network
-    );
-  } else if (chain === "near") {
-    await execute_near(parsed_vaa.payload, vaa_hex, network);
-  } else if (chain === "injective") {
-    await execute_injective(parsed_vaa.payload, buf, network);
-  } else if (chain === "xpla") {
-    await execute_xpla(parsed_vaa.payload, buf, network);
-  } else if (chain === "sei") {
-    await execute_sei(parsed_vaa.payload, buf, network);
-  } else if (chain === "osmosis") {
-    throw Error("OSMOSIS is not supported yet");
-  } else if (chain === "sui") {
-    await submitSui(parsed_vaa.payload, buf, network, argv["rpc"]);
-  } else if (chain === "aptos") {
-    await execute_aptos(
-      parsed_vaa.payload,
-      buf,
-      network,
-      argv["contract-address"],
-      argv["rpc"]
-    );
-  } else if (chain === "wormchain") {
-    throw Error("Wormchain is not supported yet");
-  } else if (chain === "btc") {
-    throw Error("btc is not supported yet");
-  } else if (chain === "sei") {
-    throw Error("sei is not supported yet");
-  } else {
-    // If you get a type error here, hover over `chain`'s type and it tells you
-    // which cases are not handled
-    impossible(chain);
+  for (let chain of chains) {
```

Comment:
Two notes about this:
1. The loop body is large, and there's some conditional behaviour based on whether the all chains flag is set to true, it would be nicer if the `execute_` logic was factored out into a separate function so it's easier to read.
2. The `for ... { await doStuff(..) }` pattern sequentialises the async operations, forcing each one to wait for the previous one. A better way is to collect all the actions into an array of promises, and call `Promise.all` https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Promise/all on them. In pseudo-ish code
```typescript
await Promise.all(chains.map((chain) => execute(chain, vaa)))
```
instead of the loop

Reply:
I refactored the code to make the for loop smaller.

Regarding using Promise.all, I considered that, but decided against it. Since the promises would be acknowledged in random order (based on the RPC responses), the output from the various chains would be comingled.

I think the most common use case for the submit command will still be for a single chain, so Promise.all wouldn't help. The most likely case when we will use `--all-chains` is when we are doing registrations for a new deployment, so the extra delay is not a big deal. In that case, I think it's more important for the output to be clear. I also don't think it's worth the additional code complexity.

So I would like to leave the code sequential. If you disagree, maybe we can discuss more off line?

Reply:
That makes sense, alternatively we could capture the output of each process and return in an array at the end to get the ordering, but it won't make too much of a difference given how infrequently this code needs to be run.